presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.scalar;
+
+import com.facebook.presto.operator.aggregation.TypedSet;
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.StandardErrorCode;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.block.BlockBuilderStatus;
+import com.facebook.presto.spi.block.InterleavedBlockBuilder;
+import com.facebook.presto.spi.type.StandardTypes;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.type.SqlType;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableList;
+
+import java.lang.invoke.MethodHandle;
+
+import static com.facebook.presto.metadata.OperatorType.CAST;
+import static com.facebook.presto.metadata.OperatorType.EQUAL;
+import static com.facebook.presto.spi.type.TypeUtils.readNativeValue;
+import static com.facebook.presto.type.TypeJsonUtils.appendToBlockBuilder;
+
+@ScalarOperator(CAST)
+public final class MapToMapCast
+{
+    private MapToMapCast() {}
+
+    @TypeParameter("FK")
+    @TypeParameter("FV")
+    @TypeParameter("TK")
+    @TypeParameter("TV")
+    @SqlType("map(TK,TV)")
+    public static Block toMap(
+            @OperatorDependency(operator = EQUAL, returnType = StandardTypes.BOOLEAN, argumentTypes = {"TK", "TK"}) MethodHandle toKeyEqualsFunction,
+            @OperatorDependency(operator = CAST, returnType = "TK", argumentTypes = {"FK"}) MethodHandle keyCastFunction,
+            @OperatorDependency(operator = CAST, returnType = "TV", argumentTypes = {"FV"}) MethodHandle valueCastFunction,
+            @TypeParameter("FK") Type fromKeyType,
+            @TypeParameter("FV") Type fromValueType,
+            @TypeParameter("TK") Type toKeyType,
+            @TypeParameter("TV") Type toValueType,
+            ConnectorSession session,
+            @SqlType("map(FK,FV)") Block fromMap)
+    {
+        // bind ConnectorSession if needed
+        if (keyCastFunction.type().parameterArray()[0] == ConnectorSession.class) {
```

Comment:
Most places allow any argument to be of type ConnectorSession (they just loop over the arguments, check for that class and skip or act accordingly). There might be a few places that expect it to be the first arg. We should clean those up.